docs/proposals/pod-resource-management.md

```diff
@@ -0,0 +1,394 @@
+      - [Top level Cgroups for QoS tiers](#top-level-cgroups-for-qos-tiers)
+      - [Pod level Cgroup creation and deletion (Docker runtime)](#pod-level-cgroup-creation-and-deletion-docker-runtime)
+      - [Container level cgroups](#container-level-cgroups)
+      - [Rkt runtime](#rkt-runtime)
+      - [Add Pod level metrics to Kubelet's metrics provider](#add-pod-level-metrics-to-kubelets-metrics-provider)
+  - [Rollout Plan](#rollout-plan)
+
+<!-- END MUNGE: GENERATED_TOC -->
+
+## Introduction
+
+As of now [Quality of Service](../../docs/design/resource-qos.md) is not enforced at a pod level. Excepting pod evictions, all the other QoS features are not applicable at the pod level.
+To better support QoS, there is a need to add support for pod level resource accounting in Kubernetes.
+
+We propose to have to have a unified cgroup hierarchy with pod level cgroups for better resource management. We will have a cgroups hierarchy with top level cgroups for the three Quality of Service(QoS) classes Guaranteed, Burstable and BestEffort. Pods (and their containers) belonging to a QoS class will be grouped under these top level QoS cgroups. And all containers in a pod are nested under the pod cgroups.
+
+The proposed cgroup hierarchy would allow for more efficient resource management and lead to improvements in node reliability.
+This would also allow for significant latency optimizations in terms of pod eviction on nodes with the use of pod level resource usage metrics.
+This document provides a basic outline of how we plan to implement and rollout this feature.
+
+
+## Non Goals
+
+- Pod level disk accounting will not be tackled in this proposal.
+- Pod level resource specification in the Kubernetes API will not be tackled in this proposal.
+
+## Motivations
+
+Kubernetes currently supports container level isolation only and lets users specify resource requests/limits on the containers [Compute Resources](../../docs/design/resources.md). Kubelet creates cgroups sandbox (via docker) for all these containers.
+
+
+There are a few shortcomings to the current model.
+ - Existing Quality of Service support does not apply to pods as a whole. On-going work to support pod level eviction using QoS requires all containers in a pod to belong to the same class. By having pod level cgroups, it is easy to track pod level usage and make eviction decisions.
+ - Infrastructure overhead per pod is currently charged to the node. The overhead of running of setting up and managing the pod sandbox is currently accounted to the node. If the pod sandbox is a bit expensive, like in the case of hyper, having pod level accounting becomes critical.
+ - For the docker runtime we have a containerd-shim which is a small library that sits in front of a runtime implementation allowing it to be repartented to init, handle reattach from the caller etc. With pod level cgroups containerd-shim can be charged to the pod instead of the machine.
```

Comment:
What course of action would you suggest. If its not fixed in docker, we would like to fix this in the kubelet. 
cc @vishh @dchen1107